WcaOnRails/app/webpacker/components/RolesTab/RoleForm.jsx

```diff
@@ -23,19 +25,58 @@ const delegateStatusOptions = ['trainee_delegate', 'candidate_delegate', 'delega
 
 export default function RoleForm({ userId, isActiveRole }) {
   const { data, loading, error } = useLoadedData(roleDataUrl(userId, isActiveRole));
+  const regionsFetching = useLoadedData(apiV0Urls.userGroups.list(groupTypes.delegate_regions));
   const { save, saving } = useSaveAction();
   const selectedGroup = groups[0].value;
   const [formValues, setFormValues] = React.useState({});
   const [apiError, setError] = React.useState(false);
   const [finished, setFinished] = React.useState(false);
 
+  const regions = React.useMemo(() => regionsFetching.data?.filter(
+    (group) => !group.parent_group_id,
+  ), [regionsFetching.data]);
+
+  const subRegions = React.useMemo(() => {
+    const subRegionsList = regionsFetching.data?.filter((group) => group.parent_group_id) || [];
+    return Object.groupBy(subRegionsList, (group) => group.parent_group_id);
+  }, [regionsFetching.data]);
+
+  React.useEffect(() => {
```

Comment:
This is a very complex effect body. Why is it necessary and are you convinced that there is no way to achieve the outcome without `useEffect`?

Reply:
This is necessary and can be done only inside useEffect because roleData.regionId can either be region ID or subregion ID. I've added more comments in the code to get more clarity on it. I couldn't find a way to do it without `useEffect`.

Reply:
(It's unfortunate that 2 things which may not be the same are both called `regionId`.)

An effect does seem necessary given the existing setup, because we need to set the initial form values once the data has loaded, and also need to allow the user to update those values via the form. If we really didn't want an effect, it would be possible to remove it by creating a separate component which is only rendered once the data has loaded, and takes the data as a parameter. That is, instead of having this after the loading/error returns
```
  return (
    <Form...
```
do this instead
```
  return (
    <LoadedRoleForm roleData={data.roleData} />
  )
```
where all the `<Form...` has been moved inside this new component `LoadedRoleForm`, along with the form state `const [formValues, setFormValues] = React.useState(calculateInitialValues(roleData));` (note the new initial state, rather than just `{}`).

Is that worth doing? :shrug: There are other nearby components that don't do it.

Reply:
I've changed the code a bit. I hope it's cleaner now. :-) Removed lots of complexities hopefully.

On a side note, if "GroupsManager with edit-features" (PR in progress) becomes successful and if the leaders like it, maybe we can bring the same for Senior Delegates as well, so that we don't have to think about any complexities like it.